builder/dockerfile/builder.go

```diff
@@ -229,6 +222,11 @@ func (b *Builder) build(source builder.Source, dockerfile *parser.Result) (*buil
 		}
 		return nil, errdefs.InvalidParameter(err)
 	}
+	for _, stage := range stages {
+		if err := system.ValidatePlatform(&stage.Platform); err != nil {
```

Comment:
This seems like a strange place to validate the platform.

Why is this necessary? Don't we already do this as the first step in `dispatch` ?

If it's still necessary then maybe `dispatchDockerfileWithCancellation` is a better place to do this validation?

Reply:
It put it here based on https://github.com/moby/moby/pull/35089#discussion_r173266198. This is the next highest place in the call-stack based on that comment, once all the dockerfile has been parsed. By validating here, it's done before any build step is executed, rather than failing after one or more build steps have been executed.

Reply:
Note this is also consistent with where all the all the instructions are validated immediately before it - again as opposed to running one or more steps, only to subsequently fail.

Reply:
That comment just says to not do it in the parser package, not necessarily to do it here.

I still don't think this is the correct place to do this validation. Stages should be validated as part of dispatch  not randomly in some other unrelated function.

Reply:
@dnephin OK, moved to dispatch